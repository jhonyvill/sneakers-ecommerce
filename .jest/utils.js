import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import theme from "../src/Themes";

export function renderWithTheme(children) {
  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
}

export const productDefault = {
  id: "id-1",
  brand: "brand mock",
  title: "title mock",
  price: 250,
  discount: 0.5,
  description: "description mock",
  pictures: [
    {
      src: "image-1.jpg",
      thumb: "thumbnail-1.jpg",
      alt: "product 1",
    },
    {
      src: "image-2.jpg",
      thumb: "thumbnail-2.jpg",
      alt: "product 2",
    },
    {
      src: "image-3.jpg",
      thumb: "thumbnail-3.jpg",
      alt: "product 3",
    },
    {
      src: "image-4.jpg",
      thumb: "thumbnail-4.jpg",
      alt: "product 4",
    },
    {
      src: "image-5.jpg",
      thumb: "thumbnail-5.jpg",
      alt: "product 5",
    },
  ],
};

export const itemDefault = {
  product: { ...productDefault },
  quantity: 1,
};
