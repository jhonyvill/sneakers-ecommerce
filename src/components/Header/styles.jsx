import styled from "styled-components";

export const Container = styled.header`
  display: grid;
  grid-template-columns: auto 1fr auto auto;
  align-items: center;
  border-bottom: 0.125rem solid ${({ theme }) => theme.colors.lightGrayishBlue};
  gap: 2.8rem;
  padding: 2rem 0;
  margin: 0 1.5rem;
  box-sizing: border-box;
  position: relative;

  @media (max-width: ${({ theme }) => theme.screen.tablet}) {
    gap: 1rem;

    & div{
      order: 2;
    }
  }
  @media (max-width: ${({ theme }) => theme.screen.largeMobile}) {
    padding: 1.5rem 0;
    border-bottom: none;
  }
  @media (max-width: ${({ theme }) => theme.screen.mobile}) {
    gap: 0.5rem;
  }
`;

export const Logo = styled.a`
  @media (max-width: ${({ theme }) => theme.screen.tablet}) {
    order: 1;
  }
  @media (max-width: ${({ theme }) => theme.screen.mobile}) {
    & svg > * {
      scale: 0.9;
    }
  }
`;

export const ProfileImage = styled.img`
  width: 3.125rem;
  min-width: 1.6rem;
  border: 1px solid transparent;
  border-radius: 50%;
  cursor: pointer;

  &:hover {
    border-color: ${({ theme }) => theme.colors.primaryOrange};
  }

  @media (max-width: ${({ theme }) => theme.screen.tablet}) {
    order: 3;
  }
  @media (max-width: ${({ theme }) => theme.screen.largeMobile}) {
    width: 1.6rem;
  }
`;
