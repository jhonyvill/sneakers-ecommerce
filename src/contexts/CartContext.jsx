import React from "react";
import { useReducer } from "react";

export const CartContext = React.createContext();

export const CartProvider = ({ children }) => {
  const initialState = { items: [] };

  function cartReducer(state, action) {
    switch (action.type) {
      case "ADD_ITEM":
        return { ...state, items: [...state.items, action.payload] };
      case "UPDATE_ITEM":
        const { product } = action.payload;
        const itemIndex = state.items.findIndex(
          (item) => item.product.id === product.id
        );

        const updatedItems = [...state.items];
        updatedItems.slice(itemIndex, product);
        return { ...state, items: updatedItems };
      default:
        state;
    }
  }

  const [state, dispatch] = useReducer(cartReducer, initialState);

  return (
    <>
      <CartContext.Provider
        value={{ cartState: state, cartDispatch: dispatch }}
      >
        {children}
      </CartContext.Provider>
    </>
  );
};
