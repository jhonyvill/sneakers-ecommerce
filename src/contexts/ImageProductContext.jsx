import React from "react";
import { useState } from "react";

export const ImageProductContext = React.createContext();

export const ImageProductProvider = ({ children }) => {
  const [selectedImage, setSelectedImage] = useState({
    image: null,
    index: 0,
  });

  const handleSelectedModalImage = (image, index) => {
    setSelectedImage({ image, index });
  };

  return (
    <>
      <ImageProductContext.Provider
        value={{
          imageSelected: selectedImage.image,
          indexImageSelected: selectedImage.index,
          handleSelectedModalImage,
        }}
      >
        {children}
      </ImageProductContext.Provider>
    </>
  );
};
