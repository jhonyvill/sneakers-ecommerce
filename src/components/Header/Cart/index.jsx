import React, { useRef, useState } from "react";
import * as S from "./styles";
import CartSvg from "../../../assets/icon-cart.svg?react";
import CartModal from "./CartModal";
import { useCart } from "../../../hooks/useCart";

const Cart = () => {
  const [isCartModalOpen, setIsCartModalOpen] = useState(false);
  const cartRef = useRef(null);
  const { totalQuantityItemsCart, cartHasItems } = useCart();

  function handleCartModal() {
    setIsCartModalOpen(!isCartModalOpen);
  }

  return (
    <div ref={cartRef}>
      <S.Cart
        aria-label="carrinho de compras"
        aria-expanded={isCartModalOpen}
        onClick={handleCartModal}
      >
        <CartSvg />
        {cartHasItems && (
          <S.CartDetail>{totalQuantityItemsCart}</S.CartDetail>
        )}
      </S.Cart>
      {isCartModalOpen && (
        <CartModal
          cartRef={cartRef}
          closeCartModal={() => setIsCartModalOpen(false)}
          hasItems={cartHasItems}
        />
      )}
    </div>
  );
};

export default Cart;
