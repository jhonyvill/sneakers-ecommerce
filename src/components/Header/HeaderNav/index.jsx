import React from "react";
import * as S from "./styles";

const HeaderNav = () => {
  return (
    <S.Container>
      <S.NavLinks>
        <S.StyledItem>
          <a href="#">Collections</a>
        </S.StyledItem>
        <S.StyledItem>
          <a href="#">Men</a>
        </S.StyledItem>
        <S.StyledItem>
          <a href="#">Women</a>
        </S.StyledItem>
        <S.StyledItem>
          <a href="#">About</a>
        </S.StyledItem>
        <S.StyledItem>
          <a href="#">Contact</a>
        </S.StyledItem>
      </S.NavLinks>
    </S.Container>
  );
};

export default HeaderNav;
