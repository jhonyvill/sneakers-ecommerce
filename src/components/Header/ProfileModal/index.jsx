import React from "react";
 import * as S from "./styles";
 
 const ProfileModal = () => {
     return(
         <S.ModalContainer>
             <S.ModalDescription>
                 Unavailable.
             </S.ModalDescription>
         </S.ModalContainer>
     );
 }

 export default ProfileModal;