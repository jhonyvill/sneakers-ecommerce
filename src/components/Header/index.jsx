import React from "react";
import * as S from "./styles";
import LogoSvg from "../../assets/logo.svg?react";
import Avatar from "../../assets/image-avatar.png";
import HeaderNav from "./HeaderNav";
import Cart from "./Cart";
import ProfileModal from "./ProfileModal";
import { useState } from "react";

const Header = () => {
  const [profileModalOpen, setProfileModalOpen] = useState(false);
  return (
    <S.Container>
      <S.Logo href="/" aria-label="Home">
        <LogoSvg data-testid="logo-svg"/>
      </S.Logo>

      <HeaderNav />
      <Cart />

      <S.ProfileImage
        src={Avatar}
        alt="imagem de perfil"
        width="100px"
        height="100px"
        onClick={() => {
          setProfileModalOpen(!profileModalOpen);
        }}
      />
      {profileModalOpen && (<ProfileModal/>)}
    </S.Container>
  );
};

export default Header;
