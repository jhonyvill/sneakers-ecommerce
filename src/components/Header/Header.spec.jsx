import { screen } from "@testing-library/react";
import { renderWithTheme } from "../../../.jest/utils";
import { CartProvider } from "../../contexts/CartContext";
import Header from "./";

function renderHeader() {
  return renderWithTheme(
    <CartProvider>
      <Header />
    </CartProvider>
  );
}

jest.mock("../../hooks/useMedia", () => ({
  __esModule: true,
  useMedia: () => {
    return false;
  },
}));

describe("<Header />", () => {
  it("should render Header component", () => {
    const { getByTestId } = renderHeader();

    expect(getByTestId("logo-svg")).toBeInTheDocument();
    expect(screen.getByRole('banner')).toBeInTheDocument();
    expect(screen.getByRole('navigation')).toBeInTheDocument();
    expect(screen.getByRole('region', {name: "carrinho de compras"})).toBeInTheDocument();
    expect(screen.getByRole('img', {name: "imagem de perfil"})).toBeInTheDocument();
  });
});

describe("<Header /> snapshots", () => {
  it("should match snapshot", () => {
    const { container } = renderHeader();
    expect(container.firstChild).toMatchSnapshot();
  });
});

