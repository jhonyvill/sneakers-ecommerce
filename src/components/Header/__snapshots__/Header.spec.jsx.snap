// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> snapshots should match snapshot 1`] = `
<header
  class="styles__Container-sc-3vm2y0-0 jxpYIj"
>
  <a
    aria-label="Home"
    class="styles__Logo-sc-3vm2y0-1 hjqGOP"
    href="#"
  >
    <icon-mock
      data-testid="logo-svg"
    />
  </a>
  <nav
    class="styles__Container-sc-ljilrx-0 bkkmpr"
  >
    <ul
      class="styles__NavLinks-sc-ljilrx-1 ehtfwm false false"
      data-testid="nav-list"
    >
      <li
        class="styles__StyledItem-sc-ljilrx-2 gRnFfe"
      >
        <a
          href="#"
        >
          Collections
        </a>
      </li>
      <li
        class="styles__StyledItem-sc-ljilrx-2 gRnFfe"
      >
        <a
          href="#"
        >
          Men
        </a>
      </li>
      <li
        class="styles__StyledItem-sc-ljilrx-2 gRnFfe"
      >
        <a
          href="#"
        >
          Women
        </a>
      </li>
      <li
        class="styles__StyledItem-sc-ljilrx-2 gRnFfe"
      >
        <a
          href="#"
        >
          About
        </a>
      </li>
      <li
        class="styles__StyledItem-sc-ljilrx-2 gRnFfe"
      >
        <a
          href="#"
        >
          Contact
        </a>
      </li>
    </ul>
  </nav>
  <section
    aria-label="carrinho de compras"
  >
    <button
      aria-expanded="false"
      class="styles__Cart-sc-33q7z1-0 cfnkFh"
    >
      <icon-mock
        data-testid="cart-svg"
      />
    </button>
  </section>
  <img
    alt="imagem de perfil"
    class="styles__ProfileImage-sc-3vm2y0-2 eNboXn"
    height="100px"
    src=""
    width="100px"
  />
</header>
`;
